firewood/src/merkle/stream.rs

```diff
@@ -88,197 +100,239 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
+                // Convert (address, index) pairs to (node, index) pairs.
+                let mut path_to_key = path_to_key
+                    .into_iter()
+                    .map(|(node, pos)| merkle.get_node(node).map(|node| (node, pos)))
+                    .collect::<Result<VecDeque<_>, _>>()
+                    .map_err(|e| api::Error::InternalError(Box::new(e)))?;
 
-                if found_key > *key {
-                    visited_node_path.pop();
+                // Tracks how much of the key has been traversed so far.
+                let mut matched_key_nibbles: Vec<u8> = vec![];
+
+                while let Some((node, pos)) = path_to_key.pop_front() {
+                    match node.inner() {
+                        NodeType::Leaf(_) => (),
+                        NodeType::Branch(branch) => {
+                            // Figure out whether we want to start iterating over this
+                            // node's children at [pos] or [pos + 1].
+                            if !path_to_key.is_empty() {
+                                // The next element in [path_to_key] will handle the child at [pos]
+                                // so we can start iterating at [pos + 1].
+                                let children_iter = get_children_iter(branch)
+                                    .filter(move |(_, child_pos)| child_pos > &pos);
+
+                                branch_iter_stack.push(BranchIterator {
+                                    key_nibbles: matched_key_nibbles.clone(),
+                                    children_iter: Box::new(children_iter),
+                                });
+                            } else {
+                                // Get the child at [pos], if any.
+                                let Some(child) = branch.children.get(pos as usize) else {
+                                    // This should never happen -- [pos] should never be OOB.
+                                    return Poll::Ready(Some(Err(api::Error::InternalError(
+                                        Box::new(api::Error::ChildNotFound),
+                                    ))));
+                                };
+
+                                let child = child
+                                    .map(|child_addr| merkle.get_node(child_addr))
+                                    .transpose()
+                                    .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                                let comparer = if child.is_none() {
+                                    // The child doesn't exist; we don't need to iterate over [pos].
+                                    |a: &u8, b: &u8| a > b
+                                } else {
+                                    // The child does exist; the first key to iterate over must be at [pos].
+                                    |a: &u8, b: &u8| a >= b
+                                };
```

Comment:
```suggestion
                                let comparer = if child.is_none() {
                                    <u8 as PartialEq>::gt
                                } else {
                                    <u8 as PartialEq>::ge
                                };
```
This probably compiles to the same thing, but you don't need a closure here; you can use a function pointer that's already defined.